narwhal/consensus/src/tests/bullshark_tests.rs

```diff
@@ -818,3 +822,257 @@ async fn restart_with_new_committee() {
         handle.await.unwrap();
     }
 }
+
+/// The test ensures the following things:
+/// * garbage collection is removing the certificates from lower rounds according to gc depth
+/// * we garbage collect/delete certificates bellow the last commit round immediately once we commit them
+/// * no certificate will ever get committed past the gc round
+/// * existing uncommitted certificates in DAG (ex from slow nodes where no-one references them) they
+/// get cleaned up.
+#[tokio::test]
+async fn garbage_collection_basic() {
+    const GC_DEPTH: Round = 4;
+    const NUM_SUB_DAGS_PER_SCHEDULE: u64 = 100;
+
+    let fixture = CommitteeFixture::builder().build();
+    let committee: Committee = fixture.committee();
+
+    // We create certificates for rounds 1 to 7. For the authorities 1 to 3 the references
+    // to previous rounds between them are fully connected - meaning that we always add all the parents
+    // from previous round, except for the authority 4 which we consider it to be slow, so no one
+    // refers to their certificates. Authority 4 still produces certificates, and uses as parents
+    // all the certificates of the other authorities but never anyone else is referring to its
+    // certificates. That will create a lone chain for authority 4. We should not see any certificate
+    // committed for authority 4.
+    let keys: Vec<PublicKey> = committee
+        .authorities()
+        .map(|(key, _)| key.clone())
+        .collect();
+    let slow_node = keys[3].clone();
+    let genesis = Certificate::genesis(&committee);
+
+    let slow_nodes = vec![(slow_node.clone(), 0.0_f64)];
+    let (certificates, _round_5_certificates) = test_utils::make_certificates_with_slow_nodes(
+        &committee,
+        1..=7,
+        genesis,
+        &keys,
+        slow_nodes.as_slice(),
+    );
+
+    // Create Bullshark consensus engine
+    let store = make_consensus_store(&test_utils::temp_dir());
+
+    let metrics = Arc::new(ConsensusMetrics::new(&Registry::new()));
+    let mut state = ConsensusState::new(metrics.clone());
+    let mut bullshark = Bullshark::new(
+        committee,
+        store,
+        GC_DEPTH,
+        metrics,
+        NUM_SUB_DAGS_PER_SCHEDULE,
+    );
+
+    // Now start feeding the certificates per round
+    for c in certificates {
+        let (_, sub_dags) = bullshark.process_certificate(&mut state, c).unwrap();
+
+        sub_dags.iter().for_each(|sub_dag| {
+            // ensure nothing has been committed for authority 4
+            assert!(
+                !sub_dag
+                    .certificates
+                    .iter()
+                    .any(|c| c.header.author == slow_node),
+                "Slow authority shouldn't be amongst the committed ones"
+            );
+
+            // Once leader of round 6 is committed then we know that garbage
+            // collection has run. In this case no certificate of round 1 should exist.
```

Comment:
Think you meant "no certificate from round 1 & 2". You have it in the assert below. The same comment applies to the tests/comments below.

Reply:
As the assert is actually the error here - we do not expect to have any certificates from round 1, as the garbage collection runs when leader 6 gets committed , with `gc_depth = 4` and calculated as `leader_round - gc_depth = 6 - 4 = 2` , and the algorithm _keeps_ everything >= 2 , so discards everything bellow that.

Reply:
Hmm I think the gc round should be the highest round that is GC'ed, not the lowest round that is not GC'ed?

Reply:
My understand of GC round comes from https://github.com/MystenLabs/sui/blob/cfac29c6f5304e311e097db7ab29bdbfd010e45c/narwhal/primary/src/core.rs#L57-L58

I believe GC round in header waiter also has the same semantics. Now this semantics is also used in CertificateFetcher, Synchronizer. I think we should change consensus to this semantics too, not the other way around. Created https://github.com/MystenLabs/sui/pull/8940